File: Assets/Greco3D/scripts/task/Control.js
#pragma strict
import System.Collections.Generic;

private var output : Output;
public var curR : int = 0;
public var curT : int =0;
public var curCity : int;
public var priorCity : int = 99;
public var trial_type : String = "null";
public var curVidNav : String;
public var coordsList = new List.<City.Coords>();

static var task_action : String;

//from control
public var curTrialList = new List.<City.Run>();
static var curStoreList = new List.<String>();

static var storeList : List.<String> = new List.<String>();

public var numCities : int;
static var city_x = new Array();
static var city_y = new Array();
private var intro : boolean = true;
private var task   : boolean = true;
private var vars: Config;

//Scripts
private var pasNav : PassiveNav;
private var timer : Timer;
private var trial : Trial;
private var videoClips: VideoClips;

//Stage gates
public var taskOn : boolean = false;
private var task_iti : boolean = true;
private var task_trial : boolean = true;
private var task_run_end : boolean = true;
//Public vars
public var get_timer: float;
public var get_task_action : String;

public var run_trial_order =new List.<int>();


function Start(){	
yield StartCoroutine(SetUpTaskType());
yield StartCoroutine(SetUpComps());
get_task_action ='run_start';
}

function Update(){

    TaskVariables();

    if(get_task_action == 'iti'){
        
	    if(task_iti){
		task_iti = false;
		task_trial = true;
	    pasNav.StopTrial();
	    trial.StartITI();
	    CityChange();
	    output.Addline();
        }
  	}
	else if(get_task_action == 'trial'){
		output.GetTrialResponse();
		if(task_trial){
		task_trial = false;
		task_iti = true;
		pasNav.StartTrial();
		trial.StopITI();
		}
  	}
	else if(get_task_action == 'run_end'){
 		if(task_run_end){
		task_run_end = false;
        RunEnd();
		}
	}

	else if(get_task_action == 'run_start'){
		get_task_action = "run_starting...";
		RunStart();						
	}
	else if(get_task_action == 'task_over'){
		get_task_action = 'task_over..';
		StartCoroutine(TaskOver());
	}

}

function RunStart(){
		run_trial_order = curTrialList[curR].trials;
//		yield StartCoroutine(trial.StartCountDown());
		yield StartCoroutine(timer.SetUpTime(vars.iti_time,vars.trial_time,vars.numT)); 
		taskOn = true;
		yield StartCoroutine(timer.StartRun());  
}

function RunEnd(){
	taskOn = false;
	yield StartCoroutine(trial.RunBreak());
	curR +=1;
	if(curR < vars.numR){
	get_task_action = 'run_start';
	taskOn = true;
	task_run_end = true;
	}
	else {

	get_task_action = 'task_over';
	}
}

function SetUpComps(){
	gameObject.AddComponent(Output); 
	yield StartCoroutine(CityConfig());
	output = GetComponent(Output);
	videoClips = GetComponent(VideoClips);
	yield;
	yield StartCoroutine(videoClips.Setup());	
}

function SetUpTaskType (){

	var config : GameObject = GameObject.Find("Config");
	vars = config.GetComponent(Config) as Config;
	 
	var player : GameObject = GameObject.Find("Player");
	pasNav = player.GetComponent(PassiveNav) as PassiveNav;
	
	timer = GetComponent(Timer) as Timer;
	trial = GetComponent(Trial) as Trial;

	//Load city stuff
	var	city : City = new City(vars.version);
	curTrialList = city.trialList;
	curStoreList = city.stores;
	coordsList = city.coordsList;
	numCities = city.coordsList.Count;
	yield;
}

function CityChange(){
    yield WaitForSeconds(0.5);


	var x =  new List.<float>();
    var y =  new List.<float>();
    var cityNum : int = run_trial_order[curT];

	x = coordsList[cityNum-1].x;
	y = coordsList[cityNum-1].y;

	//Reposition Control.curStoreList
	for( var i : int = 0; i < curStoreList.Count; i++){
        GameObject.Find(curStoreList[i]).transform.position.x = x[i];
		GameObject.Find(curStoreList[i]).transform.position.z = y[i];  			  	  			  			  
	}	
}

function TaskOver(){
	get_task_action = "task_over..";
	yield StartCoroutine(trial.TaskOver());
	var expObj : GameObject = GameObject.Find("Experiment");
	var expScript: Experiment = expObj.GetComponent("Experiment") as Experiment;
	expScript.LoadNextModule();
}

function TaskVariables(){
	if(taskOn){
	get_task_action = timer.GetAction();
	get_timer = timer.GetTime('trial');
 	curT = timer.cnt_trial;
	curCity = run_trial_order[curT];
 	curVidNav = pasNav.cityVidTrial as String;

	if(curT != 0){
	priorCity = run_trial_order[curT-1];
	}

  	if(priorCity == curCity){
       trial_type = 's';
      } else {
       trial_type = 'd';
     }
	}
}

function CityConfig(){

	var city : GameObject = GameObject.Find("City");
	yield;
	for ( var store : int = 0; store < Control.curStoreList.Count; store++){
			Control.storeList.Add(Control.curStoreList[store]);
		}
	yield;

	//Remotes stores not used in task
	for (var child : Transform in city.transform) {
	
		if(!Control.storeList.Contains(child.name)){
		
			Destroy(GameObject.Find(child.name));
	    }
	}
	yield;
}

function CheckInView(objectName : String):boolean{
	var inView : boolean;
	var store = GameObject.Find(objectName);
	var allChildren : Renderer[] =store.GetComponentsInChildren(Renderer) as Renderer[];
		for(var i : Renderer in allChildren){
			 if(i.isVisible){
			 inView =true;	 
			 }
			 else{
			 inView =false;
			 }	 
		}		
		return inView;
}

function CountInView(stores : List.<String>){
 	var cnt: int =0;
	for( var i : int = 0;i < stores.Count; i++){
	 cnt += i;
	}
	return cnt;
}

function GetInViewList(storeList:List.<String>){
	var storesInView = new List.<String>();
	for( var store : String in storeList){
	 	if(CheckInView(store)){
		storesInView.Add(store);
		}
	}
}

File: Assets/Greco3D/scripts/Greco3D/Control.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Control : MonoBehaviour {

    OutputManager output;
    int curR, curT, priorCity, cityNum, numCities;
    string trialType, curVidNav, taskAction, get_task_action;
    bool taskOn, task_iti, task_trial, task_run_end, task_start, task_run_start;
    List<City.Coords> coordsList = new List<City.Coords>();
    List<string> curStoreList = new List<string>();
    List<City.Run> runList = new List<City.Run>();
    List<int> run_trial_order = new List<int>();

    void Start () {
        StartCoroutine(StartTask());
	}
	
    IEnumerator StartTask()
    {
        taskOn = false;
        task_iti = true;
        task_trial = true;
        task_run_end = true;
        task_start = true;
        task_run_start = true;

        StartCoroutine(SetUpTaskType());
        yield return null;
        StartCoroutine(SetupComps());
        yield return null;
        get_task_action = "run_start";
    }

	void Update () {

        TaskVariables();

        if (get_task_action == "iti")
        {

            if (task_iti)
            {
                task_iti = false;
                task_trial = true;
                //pasNav.StopTrial();
                //trial.StartITI();
                NextTrialSetup();
                //output.Addline();
            }
        }
        else if (get_task_action == "trial")
        {
            //output.GetTrialResponse();
            if (task_trial)
            {
                task_trial = false;
                task_iti = true;
                //pasNav.StartTrial();
                //trial.StopITI();
            }
        }
        else if (get_task_action == "start")
        {
            if (task_start)
            {
                task_start = false;
                task_trial = true;
                //trial.StartITI();
                NextTrialSetup();
            }

        }
        else if(get_task_action == "run_end"){
         	if(task_run_end){
        	task_run_end = false;
            RunEnd();
            }
        }

        else if(get_task_action == "run_start"){
        	if(task_run_start){
        	task_run_start = false;
        	RunStart();
        	}						
        }
        else if(get_task_action == "task_over"){
        	get_task_action = "task_over..";
        	StartCoroutine(TaskOver());
        }

    }

    IEnumerator RunStart()
    {
        City.Run run = runList[curR];
        run_trial_order =run.trials;
        print("run_trial_order");

        yield return null;

        //StartCoroutine(trial.StartCountDown());

        yield return null;
        
        //StartCoroutine(timer.SetUpTime(vars.iti_time, vars.trial_time, vars.numT));
        //timer.StartRun();

        taskOn = true;
    }

    IEnumerator RunEnd()
    {
        taskOn = false;
        yield return null;
        
        //StartCoroutine(trial.RunBreak());

        curR += 1;
        if (curR < Manager.config.numR)
        {
            task_run_start = true;
            get_task_action = "run_start";
            task_run_end = true;
            task_start = true;

        }
        else
        {
            get_task_action = "task_over";
        }
    }

    IEnumerator SetupComps()
    {
        StartCoroutine(CityConfig());
        yield return null;
    }

    IEnumerator SetUpTaskType()
    {
        GameObject player = GameObject.Find("Player");
        PassiveNavC pasNav = player.GetComponent(typeof(PassiveNavC)) as PassiveNavC;

        //timer = GetComponent(typeof(Timer)) as Timer;
        //trial = GetComponent(typeof(Timer)) as Trial;

        //Load city stuff
        City city = new City(Manager.config.version);
        runList = city.runList;
        curStoreList = city.stores;
        coordsList = city.coordsList;
        numCities = city.coordsList.Count;
        yield return null;
    }

    IEnumerator NextTrialSetup()
    {
        if (curt < Manager.config.numT)
        {
            yield return WaitForSeconds(0.5);
            citynum = run_trial_order[curt] - 1;

            debug.log("setup city number" + citynum);
            pasnav.setuptrial(citynum);

            var x = new list.< float > ();
            var y = new list.< float > ();



            x = coordslist[citynum].x;
            y = coordslist[citynum].y;

            reposition control.curstorelist
            for (var i : int = 0; i < curstorelist.count; i++){
                gameobject.find(curstorelist[i]).transform.position.x = x[i];
                gameobject.find(curstorelist[i]).transform.position.z = y[i];
            }
        }
    }

    IEnumerator TaskOver()
    {
        get_task_action = "task_over..";
        yield return null;
        //StartCoroutine(trial.TaskOver());
        GameObject expObj = GameObject.Find("Experiment");
        Experiment expScript = expObj.GetComponent("Experiment") as Experiment;
        expScript.StartNextTask();
    }

    void TaskVariables()
    {
        if (taskOn)
        {
            //get_task_action = timer.getaction();
            //get_timer = timer.gettime("trial");
            //curT = timer.cnt_trial;
            //curvidnav = pasnav.cityvidtrial as string;

            if (curT != 0)
            {
                priorCity = run_trial_order[curT - 1];
            }

            if (priorCity == cityNum)
            {
                trialType = "s";
            }
            else
            {
                trialType = "d";
            }
        }
    }

    IEnumerator CityConfig()
    {

        GameObject city = GameObject.Find("City");
        //yield return null;

        //for (int store = 0; store < curStoreList.Count; store++){
        //    storeList.Add(curStoreList[store]);
        //}
        yield return null;

        //Remotes stores not used in task
        foreach (Transform child in city.transform)
        {

            if (!curStoreList.Contains(child.name))
            {

                Destroy(GameObject.Find(child.name));
            }
        }
        yield return null;
    }

    //bool CheckInView(string objectName){
    //	bool inView;
    //	GameObject store = GameObject.Find(objectName);
    //    Renderer[] allChildren = store.GetComponentsInChildren(typeof(Renderer)) as Renderer[];
    //		foreach(Renderer i in allChildren){
    //			 if(i.isVisible){
    //			 inView =true;	 
    //			 }
    //			 else{
    //			 inView =false;
    //			 }	 
    //		}		
    //	return inView;
    //}

    //int countinview(List<string> stores)
    //{
    //    int cnt = 0;
    //    for (int i = 0; i < stores.Count; i++){
    //        cnt += i;
    //    }
    //    return cnt;
    //}

    //void getinviewlist(List<string> storeList)
    //{
    //    List<string> storesInView = new List<string>();
    //    foreach (string store in storeList)
    //    {
    //        if (CheckInView(store))
    //        {
    //            storesInView.Add(store);
    //        }
    //    }
    //}

}
